src/main/scala/firrtl/passes/NewLowerTypes.scala

```diff
@@ -0,0 +1,454 @@
+          } else {
+            List(oldPortRef.field(portField.name))
+          }
+
+          val newPortType = if(isDataField) { newMem.dataType } else { portField.tpe }
+          val newPortFieldAccess = SubField(newPortAccess, portField.name, newPortType)
+
+          // record renames only for the data field which is the only port field of non-ground type
+          val newPortFieldRef = newPortRef.field(portField.name)
+          if(memWasRenamed && isDataOrMaskField) {
+            oldFieldRefs.foreach { o => renameMap.record(o, newPortFieldRef) }
+          }
+
+          val oldFieldStringRef = extractGroundTypeRefString(oldFieldRefs)
+          (oldFieldStringRef, newPortFieldAccess)
+        }
+      }
+      (newMem, subFields)
+    }
+
+    (newMemAndSubFields.map(_._1), newMemAndSubFields.flatMap(_._2))
+  }
+
+  private def memBundle(mem: DefMemory): Field = mem.dataType match {
+    case _: GroundType => Field(mem.name, Default, mem.dataType)
+    case _: BundleType | _: VectorType =>
+      val subMems = getFields(mem.dataType).map(f => mem.copy(name = f.name, dataType = f.tpe))
+      val fields = subMems.map(memBundle)
+      Field(mem.name, Default, BundleType(fields))
+  }
+
+  private def recordRenames(fieldToRefs: Seq[(Field, Seq[ReferenceTarget])], renameMap: RenameMap, parent: ParentRef):
+  Unit = {
+    // TODO: if we group by ReferenceTarget, we could reduce the number of calls to `record`. Is it worth it?
+    fieldToRefs.foreach { case(field, refs) =>
+      val fieldRef = parent.ref(field.name)
+      refs.foreach{ r => renameMap.record(r, fieldRef) }
+    }
+  }
+
+  private def extractGroundTypeRefString(refs: Seq[ReferenceTarget]): String = {
+    if (refs.isEmpty) { "" } else {
+      // Since we depend on ExpandConnects any reference we encounter will be of ground type
+      // and thus the one with the longest access path.
+      refs.reduceLeft((x, y) => if (x.component.length > y.component.length) x else y)
+        // convert references to strings relative to the module
+        .serialize.dropWhile(_ != '>').tail
+    }
+  }
+
+  private def destruct(m: ModuleTarget, field: Field, rename: Option[RenameNode]): Seq[(Field, Seq[ReferenceTarget])] =
+    destruct(prefix = "", oldParent = ModuleParentRef(m), oldField = field, isVecField = false, rename = rename)
+
+  /** Lowers a field into its ground type fields.
+    * @param prefix carries the prefix of the new ground type name
+    * @param isVecField is used to generate an appropriate old (field/index) reference
+    * @param rename The information from the `uniquify` function is consumed to appropriately rename generated fields.
+    * @return a sequence of ground type fields with new names and, for each field,
+    *         a sequence of old references that should to be renamed to point to the particular field
+    */
+  private def destruct(prefix: String, oldParent: ParentRef, oldField: Field,
```

Comment:
I'd add a comment that this function presumes the fields have already been uniquified.

Reply:
The fields haven't really been uniquified quite yet. The uniquifying information has been calculated, but we actually apply it in this function (which is what the `rename` parameter is for).
Is there a way I could clarify that?